test/params_test.c

```diff
@@ -0,0 +1,145 @@
+/*
+ * Copyright 2019 The OpenSSL Project Authors. All Rights Reserved.
+ * Copyright (c) 2019, Oracle and/or its affiliates.  All rights reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <string.h>
+#include "testutil.h"
+#include "internal/nelem.h"
+#include <openssl/params.h>
+#include <openssl/bn.h>
+
+#define MAX_LEN 20
+
+static void swap_copy(unsigned char *out, const void *in, size_t len)
+{
+    size_t j;
+
+    for (j = 0; j < len; j++)
+        out[j] = ((unsigned char *)in)[len - j - 1];
+}
+
+static void copy_to_le(unsigned char *out, const void *in, size_t len)
+{
+#ifdef B_ENDIAN
+    swap_copy(out, in, len);
+#else
+    memcpy(out, in, len);
+#endif
+}
+
+static void copy_be_to_native(unsigned char *out, const void *in, size_t len)
+{
+#ifdef B_ENDIAN
+    memcpy(out, in, len);
+#else
+    swap_copy(out, in, len);
+#endif
+}
+
+static const struct {
+    size_t len;
+    unsigned char value[MAX_LEN];
+} raw_values[] = {
+    { 4, { 0x38, 0x27, 0xbf, 0x3b } },
+    { 8, { 0x59, 0xb2, 0x1a, 0xe9, 0x2a, 0xd8, 0x46, 0x40 } },
+};
+
+static int test_param_type_extra(const OSSL_PARAM *param,
+                                 const unsigned char *cmp)
+{
+    int32_t i32;
+    double d;
+    int64_t i64;
+    size_t s;
+    unsigned char buf[MAX_LEN];
+    const int bit32 = param->buffer_size == sizeof(int32_t);
+
+    if ((bit32 && !TEST_true(OSSL_PARAM_get_int32(param, "a", &i32)))
+        || !TEST_true(OSSL_PARAM_get_int64(param, "a", &i64))
+        || !TEST_true(OSSL_PARAM_get_size_t(param, "a", &s))
+        || !TEST_true(OSSL_PARAM_get_double(param, "a", &d)))
+        return 0;
+
+    if (bit32) {
+        copy_to_le(buf, &i32, sizeof(i32));
+        if (!TEST_mem_eq(buf, sizeof(i32), cmp, sizeof(i32)))
+            return 0;
+    }
+    copy_to_le(buf, &i64, sizeof(i64));
+    if (!TEST_mem_eq(buf, sizeof(i64), cmp, sizeof(i64)))
+        return 0;
+    copy_to_le(buf, &s, sizeof(s));
+    if (!TEST_mem_eq(buf, sizeof(s), cmp, sizeof(s)))
+        return 0;
+
+    /* reals lose accuracy */
+    if (param->buffer_size <= 6) {
+        i64 = (int64_t)d;
+        copy_to_le(buf, &i64, sizeof(i64));
+        if (!TEST_mem_eq(buf, param->buffer_size, cmp, param->buffer_size))
+            return 0;
+    }
+     return 1;
+}
+
+#define TEST_INT(name, type, neg) \
```

Comment:
Could you add some commentary about the purpose of this macro and what its code is attempting to do?  I *think* I got it, but it wasn't an easy read.

Reply:
Done.  Is that sufficient?

Reply:
That's perfect.  That last sentence was illuminating, *now* I fully understand why you're doing all that work.